packages/protocol/contracts/governance/VestingInstance.sol

```diff
@@ -0,0 +1,332 @@
+   * @param vestingRefundDestination address of the refund receiver after the vesting is deemed revoked
+   * @param registryAddress address of the deployed contracts registry
+   */
+  constructor(
+    address vestingBeneficiary,
+    uint256 vestingAmount,
+    uint256 vestingCliff,
+    uint256 vestingStartTime,
+    uint256 vestingPeriodSec,
+    uint256 vestAmountPerPeriod,
+    bool vestingRevocable,
+    address vestingRevoker,
+    address vestingRefundDestination,
+    address registryAddress
+  ) public {
+    require(vestingAmount > 0, "Vesting amount must be positive");
+    require(vestingBeneficiary != address(0), "Beneficiary is the zero address");
+    require(vestingRefundDestination != address(0), "Refund destination is the zero address");
+    require(vestingCliff <= vestingPeriodSec, "Vesting cliff is longer than vesting duration");
+    require(vestingPeriodSec > 0, "Vesting period is 0 s.");
+    require(
+      vestAmountPerPeriod <= vestingAmount,
+      "Vesting amount per period is greater than the total vesting amount"
+    );
+    require(
+      vestingAmount.div(vestAmountPerPeriod) >= 1,
+      "There must be more than 1 vesting periods"
+    );
+    require(
+      vestingStartTime.add(vestingCliff) > block.timestamp,
+      "Vesting end time is before current time"
+    );
+
+    setRegistry(registryAddress);
+
+    vestingSchedule.vestingAmount = vestingAmount;
+    vestingSchedule.vestAmountPerPeriod = vestAmountPerPeriod;
+    vestingSchedule.vestingPeriodSec = vestingPeriodSec;
+    vestingSchedule.vestingCliffStartTime = vestingStartTime.add(vestingCliff);
+    vestingSchedule.vestingStartTime = vestingStartTime;
+
+    beneficiary = vestingBeneficiary;
+    revocable = vestingRevocable;
+    refundDestination = vestingRefundDestination;
+    revoker = vestingRevoker;
+  }
+
+  /**
+   * @notice Transfers available released tokens from the vesting back to beneficiary.
+   */
+  function withdraw() external nonReentrant onlyBeneficiary {
+    bool isUnpaused = !paused || (paused && block.timestamp >= pauseEndTime);
+    require(isUnpaused, "Withdrawals only allowed in the unpaused state");
+    if (isUnpaused) paused = false;
+    uint256 withdrawableAmount = getWithdrawableAmountAtTimestamp(block.timestamp);
+    require(withdrawableAmount > 0, "No gold is due for withdrawal");
+    currentlyWithdrawn = currentlyWithdrawn.add(withdrawableAmount);
+    require(getGoldToken().transfer(beneficiary, withdrawableAmount), "Withdrawal of gold failed");
+  }
+
+  /**
+   * @notice Allows only the revoker to revoke the vesting. Gold already vested
+   * remains in the contract, the rest is returned to the _refundDestination.
+   * @param revokeTimestamp the revocation timestamp
+   * @dev If revokeTimestamp is less than the current block timestamp, it is set equal to the latter
+   */
+  function revoke(uint256 revokeTimestamp) external nonReentrant onlyRevoker {
```

Comment:
Locked gold rewards (& penalties) are unkown for a future revokeTimestamp... so it's impossible to know what belong to the revoker and what to the beneficiary

Reply:
Which makes me wonder... why do we need to pass a timestamp.. should revoke be "revoke now"?